packages/parse5-parser-stream/lib/index.ts

```diff
@@ -0,0 +1,105 @@
+import { Writable } from 'node:stream';
+import { Parser, ParserOptions } from 'parse5/dist/parser/index.js';
+import type { TreeAdapterTypeMap } from 'parse5/dist/tree-adapters/interface.js';
+import type { DefaultTreeAdapterMap } from 'parse5/dist/tree-adapters/default.js';
+
+/**
+ * Streaming HTML parser with scripting support.
+ * A [writable stream](https://nodejs.org/api/stream.html#stream_class_stream_writable).
+ *
+ * @example
+ *
+ * ```js
+ * const ParserStream = require('parse5-parser-stream');
+ * const http = require('http');
+ *
+ * // Fetch the page content and obtain it's <head> node
+ * http.get('http://inikulin.github.io/parse5/', res => {
+ *     const parser = new ParserStream();
+ *
+ *     parser.once('finish', () => {
+ *         console.log(parser.document.childNodes[1].childNodes[0].tagName); //> 'head'
+ *     });
+ *
+ *     res.pipe(parser);
+ * });
+ * ```
+ *
+ */
+export class ParserStream<T extends TreeAdapterTypeMap = DefaultTreeAdapterMap> extends Writable {
+    private lastChunkWritten = false;
+    private writeCallback: null | (() => void) = null;
+    private pausedByScript = false;
+
+    public parser: Parser<T>;
+    private pendingHtmlInsertions: string[] = [];
+    /** The resulting document node. */
+    public document: T['document'];
+
+    constructor(options?: ParserOptions<T>) {
+        super({ decodeStrings: false });
+
+        this.parser = new Parser(options);
+
+        this.document = this.parser.treeAdapter.createDocument();
+        this.parser._bootstrap(this.document, null);
+    }
+
+    //WritableStream implementation
+    override _write(chunk: string, _encoding: string, callback: () => void): void {
+        if (typeof chunk !== 'string') {
+            throw new TypeError('Parser can work only with string streams.');
+        }
+
+        this.writeCallback = callback;
+        this.parser.tokenizer.write(chunk, this.lastChunkWritten);
+        this._runParsingLoop();
+    }
+
+    // TODO [engine:node@>=16]: Due to issues with Node < 16, we are overriding `end` instead of `_final`.
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    override end(chunk?: any, encoding?: any, callback?: any): any {
+        this.lastChunkWritten = true;
+        super.end(chunk || '', encoding, callback);
+    }
+
+    //Scriptable parser implementation
+    private _runParsingLoop(): void {
+        this.parser.runParsingLoopForCurrentChunk(this.writeCallback, this._scriptHandler);
+    }
+
+    private _resume = (): void => {
```

Comment:
don't suppose you know why these are all properties rather than methods? we should be able to pass a method as a script handler since it still matches the signature

Reply:
If we passed methods below, we wouldn't get a binding of the `this` value. Using arrow functions works around that.